client/terraform_cli.go

```diff
@@ -194,6 +195,45 @@ func (t *TerraformCLI) Plan(ctx context.Context) (bool, error) {
 	return t.tf.Plan(ctx, opts...)
 }
 
+// Validate verifies the generated configuration files
+func (t *TerraformCLI) Validate(ctx context.Context) error {
+	output, err := t.tf.Validate(ctx)
+	if err != nil {
+		return err
+	}
+
+	var sb strings.Builder
+	for _, d := range output.Diagnostics {
+		sb.WriteByte('\n')
+		switch d.Detail {
+		case `An argument named "services" is not expected here.`:
+			fmt.Fprintf(&sb, "%s is missing the \"services\" variable\n", *d.Snippet.Context)
+		case `An argument named "catalog_services" is not expected here.`:
```

Comment:
I don't think the error verbiage around this has changed much at all across Terraform versions, so this would be fine. Just wanted to it call-out in case we stop seeing this useful new error, we could suspect that the error string has changed.

Reply:
I could add an integration or e2e tests for these two cases specifically so that when we test new versions of Terraform, we'd be able to tell if an error message changed.

Reply:
Good idea about an integration/e2e test! Doesn't have to block this PR and a follow-up would work too.

Another thought came to mind, do you know how this edge-case handling would catch if the `var.services` type declared in the module is not compatible with the type CTS expects? not sure how to clearly say that, so here's an example:

[CTS expects this type](https://www.consul.io/docs/nia/terraform-modules#services-variable)
```hcl
variable "services" {
  type = map(
    object({
      id          = string
      name      = string
      ...
    })
  )
}
```

Terraform module used has a different type for declared in its `variables.tf` file
```hcl
variable "services" {
  type = list(string)
}
```

Reply:
Here's the current output:

```
2021/06/10 16:23:13.736550 [ERR] (cli) error initializing controller: 
error: Invalid value for module argument

on main.tf line 31, in module "example-task"
31:  services         = var.services

The given value is not suitable for child module variable "services" defined at .terraform/modules/example-task/variables.tf:37,1-20: list of string required.
```

I could also check for this error with services and catalog_services.

Reply:
Thanks for checking, the error seems informative enough for me!